contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md

```diff
@@ -0,0 +1,280 @@
+to provide high throughput at the cost of being simple.
+To help in building a scheduler (the default or a custom one) and to share
+elements of the scheduling logic,
+[the scheduling framework](https://www.google.com/url?q=https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/&sa=D&source=editors&ust=1617027676318000&usg=AOvVaw2vEsCZwet-BiQGhy0mVpZu)
+was implemented.
+The framework does not provide all pieces to build a new scheduler from scratch.
+Queues, caches, scheduling algorithms and other building elements are still needed to assemble
+a fully functional unit. This document aims at describing how all the individual
+pieces are put together and what’s their role in the overall architecture
+so a developer can quickly orient in the code.
+
+## Scheduling a pod
+
+The default scheduler instance has a loop running indefinitely
+which (everytime there’s a pod) is responsible for invoking the scheduling logic
+and making sure a pod gets either a node assigned or requeued for future processing.
+Each loop consists of a scheduling and a binding cycle.
+The scheduling cycle is responsible for running the scheduling algorithm selecting
+the most suitable node for placing the pod.
+The binding cycle makes sure the kube-apiserver is made aware of the selected
+node at the right time (e.g. gang scheduling requires to wait until all pods
+in the gang have their node selected, binding all pods at once).
+
+### Scheduling cycle
+
+Each cycle honors the following steps:
+1. Get the next pod for scheduling
+1. Schedule a pod with provided algorithm
+1. If a pod fails to be scheduled due to `FitError`, run preemption plugin in
+   `PostFilterPlugin` (if the plugin is registered) to nominate a node where
+   the pods can run. If preemption was successful,
+   let the current pod be aware of the nominated node.
+   Handle the error, get the next pod and start over.
+1. If the scheduling algorithm finds a suitable node, store the pod into
+   the scheduler cache (`AssumePod` operation) and run plugins from the `Reserve`
+   and `Permit` extension point in that order. In case either of the plugins fails,
+   end the current scheduling cycle, increase relevant metrics and handle
+   the scheduling error through the `Error` handler.
+1. Upon successfully running all extension points, proceed to the binding cycle.
+   At the same time start processing another pod (if there’s any).
+
+### Binding cycle
+
+Consists of the following four extension points ran in the same order:
+- [WaitOnPermit](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L560)
+  (internal API): Some plugins from the Permit extension point
+  may send a request for an operation requiring to wait for a condition
+  (e.g. wait for additional resources to be available or wait for all pods
+  in a gang to be assumed).
+  The `WaitOnPermit` extension point allows you to wait for such a condition to be met.
```

Comment:
Generally `WaitOnPermit` is the 2nd part of `Permit` extension point, and works internally on the framework side. So maybe reword it a bit to avoid confusion as when users think of "extension point",  they usually think of it as an extendable interface, which is `Permit()`.

```
Under the hood, `WaitOnPermit` waits for such a condition to be met within a timeout threshold.
```